### PR TITLE
Refactored buffer and buffer_view

### LINKED ISSUE #45: Refactor buffer internal representation
The buffer class should use 3 pointers instead of pointer + size (+ capacity), as the implementation of `std::vector` does.
This should improve performance and make it easier to implement the insert / reserve / push_back methods, as well as the noexcept handling.

### PR DESCRIPTION
**Pull request by JohanMabille**
Fix #45

### CODE REVIEW

File: include/sparrow/buffer.hpp [hunk 1]
```diff
@@ -18,454 +18,870 @@
...
     }
 
     template <class T>
-    bool buffer_base<T>::empty() const noexcept
+    buffer<T>& buffer<T>::operator=(const buffer& rhs)
     {
-        return size() == size_type(0);
+        if (std::addressof(rhs) != this)
+        {
+            // We assume that any_allocator nevers propagates on assign
```
Comment (jjerphan) [inline include/sparrow/buffer.hpp:500]: What do you mean by this comment exactly?
```suggestion
            // We assume that any_allocator never propagates on assign.
```

File: include/sparrow/buffer.hpp [hunk 2]
```diff
@@ -18,454 +18,870 @@
...
     }
 
     template <class T>
-    buffer<T>::buffer(pointer data, size_type size)
-        : base_type{data, size}
+    constexpr void buffer<T>::reserve(size_type new_cap)
     {
+        if (new_cap > max_size())
+        {
+            throw std::length_error("buffer::reserve called with new_cap > max_size()");
```
Comment (jjerphan) [inline include/sparrow/buffer.hpp:697]: What should the policy for raising exception be?

File: include/sparrow/buffer.hpp [hunk 3]
```diff
@@ -18,454 +18,870 @@
...
     }
 
     template <class T>
-    void buffer<T>::clear()
+    constexpr auto buffer<T>::max_size_impl(const allocator_type& a) noexcept -> size_type
     {
-        resize(size_type(0));
+        const size_type diff_max = std::numeric_limits<difference_type>::max();
+        const size_type alloc_max = std::allocator_traits<allocator_type>::max_size(a);
+        return (std::min)(diff_max, alloc_max);
```
Comment (jjerphan) [inline include/sparrow/buffer.hpp:836]: Why is this needed?
```suggestion
        return std::min(diff_max, alloc_max);
```

**Review by jjerphan — COMMENTED** [review]
A first brief pass.

File: include/sparrow/buffer.hpp (hunk 1 above)
Comment (JohanMabille) [reply include/sparrow/buffer.hpp:500]: We do not assign allocators from a buffer instance to another one.

File: include/sparrow/buffer.hpp (hunk 2 above)
Comment (JohanMabille) [reply include/sparrow/buffer.hpp:697]: I took the same as `std::vector`, see the `Exceptions` section [here](https://en.cppreference.com/w/cpp/container/vector/vector) and [here](https://en.cppreference.com/w/cpp/container/vector/reserve)

File: include/sparrow/buffer.hpp (hunk 3 above)
Comment (JohanMabille) [reply include/sparrow/buffer.hpp:836]: Because of the min macro defined on Windows, but we can remove it if we define `NOMINMAX` and propagate it.

File: include/sparrow/buffer.hpp [hunk 4]
```diff
@@ -18,454 +18,870 @@
...
+        m_data.p_storage_end = m_data.p_begin + n;
+    }
 
-        buffer_view(buffer_view&&) = default;
-        buffer_view& operator=(buffer_view&&) = default;
-    };
+    template <class T>
+    constexpr void buffer_base<T>::assign_storage(pointer p, size_type n, size_type cap)
+    {
+        m_data.p_begin = p;
```
Comment (jjerphan) [inline include/sparrow/buffer.hpp:379]:
```suggestion
        assert(n <= cap);
        m_data.p_begin = p;
```

File: include/sparrow/buffer.hpp [hunk 5]
```diff
@@ -18,454 +18,870 @@
...
+            constexpr buffer_data& operator=(buffer_data&&) noexcept;
+        };
+
+        buffer_base() = default;
+        
+        template <allocator A>
+        constexpr buffer_base(const A& a) noexcept;
+
+        template <allocator A = allocator_type>
+        buffer_base(size_type n, const A& a = A());
```
[the remark this answers is not in the document]
Comment (jjerphan) [reply include/sparrow/buffer.hpp:66]: Are there any reasons for it not to be `constexpr`?

File: include/sparrow/buffer.hpp
Comment (jjerphan) [inline include/sparrow/buffer.hpp:1]: This file has to be clang-formatted.

File: include/sparrow/buffer.hpp [hunk 6]
```diff
@@ -18,454 +18,870 @@
...
     }
 
     template <class T>
-    void buffer<T>::resize(size_type n)
+    template <class It>
+    constexpr auto buffer<T>::allocate_and_copy(size_type n, It first, It last) -> pointer
     {
-        // TODO: add capacity, resize if growing only and define a shrink_to_fit method
-        if (n != base_type::size())
+        pointer p = this->allocate(n);
```
Comment (jjerphan) [inline include/sparrow/buffer.hpp:808]: Must it be present in the exception handling bellow?

File: include/sparrow/buffer.hpp [hunk 7]
```diff
@@ -18,454 +18,870 @@
...
+        try
         {
-            buffer<T> tmp(n);
-            const size_type copy_size = std::min(base_type::size(), n);
-            std::copy(base_type::data(), base_type::data() + copy_size, tmp.data());
-            base_type::swap(tmp);
+            copy_initialize(first, last, p, get_allocator());
         }
+        catch(...)
```
Comment (jjerphan) [inline include/sparrow/buffer.hpp:813]: If l.808 is exception-handled, this this be more specific?

**Review by jjerphan — COMMENTED** [review]
LGTM modulo a few last comments.

File: include/sparrow/buffer_view.hpp [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+// limitations under the License.
+
+#pragma once
+
+#include "sparrow/buffer.hpp"
+
+namespace sparrow
+{
+    /*
+     * @class buffer_view
```
Comment (Klaim) [inline include/sparrow/buffer_view.hpp:22]: You can remove this and rely on the autobrief syntax?

File: include/sparrow/buffer_view.hpp [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+        using size_type = std::size_t;
+        using difference_type = std::ptrdiff_t;
+
+        using iterator = pointer_iterator<pointer>;
+        using const_iterator = pointer_iterator<const_pointer>;
+        using reverse_iterator = std::reverse_iterator<iterator>;
+        using const_reverse_iterator = std::reverse_iterator<const_iterator>;
+
+        explicit buffer_view(buffer<T>& buffer);
+        buffer_view(pointer p, size_type n);
```
Comment (Klaim) [inline include/sparrow/buffer_view.hpp:45]: To be complete we also need a `subrange(...)` function and a constructor that takes a pair of iterators (they should be template parameters constrained by concept, not `buffer_view::iterator`).
These are often used with view types (see `std::span`, which also provides `first` and `last` subview functions).

File: include/sparrow/buffer_view.hpp [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+        reverse_iterator rbegin();
+        reverse_iterator rend();
+
+        const_reverse_iterator rbegin() const;
+        const_reverse_iterator rend() const;
+        const_reverse_iterator crbegin() const;
+        const_reverse_iterator crend() const;
+
+        void swap(buffer_view& rhs) noexcept;
+        bool equal(const buffer_view& rhs) const;
```
Comment (Klaim) [inline include/sparrow/buffer_view.hpp:82]: This is not standard but is public, could you document what it does exactly? It compares elements? Is it necessary when we have `std::equal` and `std::ranges::equal` already? Or did you intend it to be private?

File: include/sparrow/buffer_view.hpp [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+
+    private:
+
+        pointer p_data = nullptr;
+        size_type m_size = 0u;
+    };
+
+    template <class T>
+    bool operator==(const buffer_view<T>& lhs, const buffer_view<T>& rhs);
```
Comment (Klaim) [inline include/sparrow/buffer_view.hpp:91]: Why not declare it in `buffer_view` as friend?

BTW should we enable comparison with another type of (contiguous?) range?

File: include/sparrow/buffer_view.hpp [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+#include "sparrow/buffer.hpp"
+
+namespace sparrow
+{
+    /*
+     * @class buffer_view
+     * @brief Object that references but does not own a piece of contiguous memory
+     */
+    template <class T>
+    class buffer_view
```
Comment (Klaim) [inline include/sparrow/buffer_view.hpp:26]: How is this type different from `std::span`?

File: test/test_buffer.cpp [hunk 13]
```diff
@@ -161,32 +168,86 @@ namespace sparrow
             CHECK_EQ(b3.data()[idx], expected_value);
         }
 
-        TEST_CASE("equality comparison")
+        // Iterators
+
```
Comment (Klaim) [inline test/test_buffer.cpp:172]: Missing tests: static check that
- for a given set of T, buffer_view<T> is a contiguous range
- buffer_view<T>::iterator is the appropriate type of iterator

File: test/test_buffer.cpp [hunk 14]
```diff
@@ -161,32 +168,86 @@ namespace sparrow
...
+            const std::size_t size = 8u;
+            buffer_test_type b(make_test_buffer(size), size);
+            auto iter = b.rbegin();
+            auto citer = b.crbegin();
+            for (std::size_t i = b.size(); i != 0u; --i)
+            {
+                CHECK_EQ(*iter++, b[i - 1]);
+                CHECK_EQ(*citer++, b[i - 1]);
+            }
+            CHECK_EQ(iter, b.rend());
```
Comment (Klaim) [inline test/test_buffer.cpp:199]: these should also match some iterators obtained through `begin/end`

File: test/test_buffer.cpp (hunk 13 above)
Comment (Klaim) [reply test/test_buffer.cpp:172]: Actually, I dont see any test specific to buffer_view?

File: include/sparrow/dynamic_bitset.hpp [hunk 15]
```diff
@@ -20,6 +20,7 @@
 #include <ranges>
 
 #include "sparrow/buffer.hpp"
+#include "sparrow/buffer_view.hpp"
```
Comment (Klaim) [inline include/sparrow/dynamic_bitset.hpp:23]: I dont see it used here?

File: include/sparrow/buffer_view.hpp (hunk 11 above)
Comment (Klaim) [reply include/sparrow/buffer_view.hpp:91]: `operator!=` is missing?

File: include/sparrow/buffer.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (JohanMabille) [reply include/sparrow/buffer.hpp:66]: Nope, I thought I fixed it Oo

File: include/sparrow/buffer_view.hpp (hunk 10 above)
Comment (JohanMabille) [reply include/sparrow/buffer_view.hpp:82]: This was to avoid declaring `operator==` as friend, but I agree this is not standard and should be removed.

File: include/sparrow/buffer.hpp (hunk 6 above)
Comment (JohanMabille) [reply include/sparrow/buffer.hpp:808]: No, the exception handling is to make sure the memory is deallocated if the initialization throws. If the allcoation throws, we have nothing to deallocate and we let the user decide what to do.

File: include/sparrow/buffer_view.hpp (hunk 11 above)
Comment (JohanMabille) [reply include/sparrow/buffer_view.hpp:91]: `operator!=` is now (C++20) automatically generated when you provide `operator==`

Comment (jjerphan) [reply include/sparrow/buffer_view.hpp:91]: > `operator!=` is missing?

In C++20, it is automatically defined after `operator==`'s definition.

File: include/sparrow/buffer_view.hpp (hunk 8 above)
Comment (JohanMabille) [reply include/sparrow/buffer_view.hpp:22]: Not sure, I will set up a skeleton for the doc and check whether we can do it in a dedicated PR. If so, I will remove all of them.

File: include/sparrow/buffer_view.hpp (hunk 9 above)
Comment (JohanMabille) [reply include/sparrow/buffer_view.hpp:45]: Maybe we can do that in a dedicated PR? This piece of code was just moved from the initial `buffer.hpp` file. I'll add a todo for that.

File: include/sparrow/buffer_view.hpp (hunk 12 above)
Comment (JohanMabille) [reply include/sparrow/buffer_view.hpp:26]: They are very similar indeed. However, `std::span` is missing some APIs in C++20 (like `cbegin` / `cend` ), and we may want to provide additional std;:vector like APIs.

File: test/test_buffer.cpp (hunk 13 above)
Comment (JohanMabille) [reply test/test_buffer.cpp:172]: These are the tests of the `buffer` class; the test suite of `buffer_view` now starts at line 315 and remains unchanged.

File: include/sparrow/dynamic_bitset.hpp (hunk 15 above)
Comment (JohanMabille) [reply include/sparrow/dynamic_bitset.hpp:23]: In the `dynamic_bitset_view` class.

**Review by jjerphan — APPROVED** [review]
LGTM up to file formatting.

File: include/sparrow/buffer_view.hpp (hunk 9 above)
Comment (Klaim) [reply include/sparrow/buffer_view.hpp:45]: I feel that this and the iterator pair constructor (which should be what the subrange is built with) not being there makes it incomplete. But ok, it can be added in another PR although please do it immediately after this one then?

File: include/sparrow/buffer_view.hpp (hunk 12 above)
Comment (Klaim) [reply include/sparrow/buffer_view.hpp:26]: Could you add that clarification in the comments, so that we keep a trace in the code?

**Review by Klaim — APPROVED** [review]
Other than the details I pointed LGTM
